### ISSUE TITLE
No need for 10ms timeslice in MediaRecorder sample

### ISSUE THREAD

**Issue by Pehrsons** [webrtc/samples#1153]
From https://github.com/webrtc/samples/blob/gh-pages/src/content/getusermedia/record/js/main.js#L109

What's the rationale behind a 10ms timeslice?

This use of the timeslice, namely storing the blobs in memory, is functionally equivalent to not having a timeslice at all.

Not having a timeslice also adds some benefits. It lets the UA optimize how it gathers data into the blobs, e.g., by storing the data in a blob on disk instead of in memory. It also simplifies the code since there's no need for a blob array or the bookkeeping that comes with it.

**Comment by fippo** [comment]
It *might* have come from the desire to send the data over a websocket? I've certainly seen it used like that, 10ms will basically try to send one slice for every frame.

Showing a longer timeslice (e.g. 30 seconds) might be something to hint at (because it allows saving a chunk in case the user closes the page, etc) but whether that makes much sense for the sample...

**Comment by Pehrsons** [comment]
Then the sample should do those things, or it's all set up for confusion, since we expect web developers to look at this and think of it as a best practice.

**Comment by ErikHellman** [comment]
Is there a tradeoff on performance and/or memory usage that we should explain? @fippo gives a good reason for using at least some timeslices, but can there be other factors that would motivate using a certain timeslice?

**Comment by Pehrsons** [comment]
If you do a really long recording with a short timeslice and store it in memory (like the sample) you're bound to run out of memory. With a sufficiently long (impossible to say what would suffice, so best to do completely without) timeslice each slice could be backed by a file and you can avoid running out of memory. For storing like this for a download, using no timeslice will always be best.

Most interesting real use cases I've seen come into play when streaming to a server or when storing locally and persistently like fippo said. In either case you have no OOM problem to consider, since you're throwing the blobs at the network or disk and then forgetting about them.

For network and storage use cases it makes sense to use a timeslice. Then you have to consider perf (this is done on main thread and can cause jank) and latency (network: well, latency; storage: how recent data should it be guaranteed to contain if we crash at any given time).

Something I've seen people do a couple of times that would speak in favor of timeslice when not involving network or storage is feeding the recorded blobs to MSE and a media element. The timeslice interval then becomes a latency question. But as I understand it such use cases are artificial demos of how broadcasting a MediaStream over a websocket could work.

TL;DR there's a tradeoff to consider when streaming/consuming the blobs as soon as they become available, but not for storing in memory.

**Comment by ErikHellman** [comment]
Thanks! I'll try to update the sample accordingly.

**Comment by Pehrsons** [comment]
@ErikHellman did you make any progress on this?

**Comment by hem1232** [comment]
Hello,

I am doing recording with media recorder and i have 50 milisec timeslice for ondataavailable method.
when we do a long recording then sometimes browser gets hang.
can anyone help me on this.

Thanks,
Hemant

**Comment by alvestrand** [comment]
Browser hangs are best solved by filing a bug on that browser - that process will usually lead you through the details of providing critical information like the browser version.

**Comment by hem1232** [comment]
i am using the latest version of chrome 80.0.

i just want to know that,record a single chunk at onstop method is good or push chunks into array after 50 milisec.

i have 50 ms timeslice for ondataavailable.

so is this the reason for browser hang?

**Comment by alvestrand** [comment]
again, please file a bug on crbug.com with the details of your setup, and how long it takes to hang your browser. If you have a script that will reproduce the hang, please attach it.
This has to be addressed by the browser developers, not by the spec sample developers.
